### DISCUSSION TITLE
如何调用其他target来生成中间文件？

### DISCUSSION THREAD

**Discussion by Domain** [xmake-io/xmake#2429] [Q&A]
我有一个target叫app，一个target叫builder，app需要builder运行生成一些源文件才能编译，该怎么写xmake.lua呢？我想在app的xmake.lua里面用`add_deps("builder")`增加依赖，然后在`before_build`里面用os.execv来调用builder，但是这里面我只能获取当前target的属性，否则我就只能硬编码builder的路径。另外，现在before_build是先执行的，这个时候builder还没生成，也会失败。

**Comment by Domain** [comment]
参考一下，fastbuild的Exec函数是可以做到的，它的ExecExecutable参数可以指定另一个目标，不用指定路径，这个目标也会先编译。我只需要把app依赖于这个Exec目标就可以了，希望xmake也可以实现类似的

**Reply by waruqi** [reply]
设置 always_added 到 add_files

 https://github.com/xmake-io/xmake/issues/1090#issuecomment-889576769

然后禁用跨 target 并行编译，https://xmake.io/#/zh-cn/manual/project_target?id=buildacross_targets_in_parallel

**Reply by Domain** [reply]
> 设置 always_added 到 add_files

这个不行，因为builder会生成很多文件，我并不能提前知道文件名

**Comment by Domain** [comment]
具体是这样的，我有很多xml文件，builder的源码，还有多个app，我需要先编译出来builder，然后builder根据这些xml生成各个app需要的源码，再编译各个app，builder只需要运行一次，而且只要builder源码不变或者xml内容不变，都不需要再次运行。
我现在想法是，定义builder这个target生成工具，自定义rule处理xml文件，增加一个run_builder的target来生成文件（我还不知道这个target的kind应该是什么？），各个app通过add_deps依赖于run_builder。
现在问题是自定义的rule里面需要调用builder来生成文件，这里面怎么获取builder的targetdir和filename？毕竟硬编码进去不太好。各个app需要包含未生成的代码，目前好像还不支持。

**Reply by waruqi** [reply]
可以看下，https://github.com/xmake-io/xmake-repo/blob/master/packages/l/luajit/port/xmake.lua 也是先编译 host bin，根据 host bin 生成后继续编译

**Reply by Domain** [reply]
谢谢！看了这个确实有所启发，已经接近解决了，不过对于across_targets_in_parallel我有个疑问，我有下面的配置：
```lua
target("common1")
    ...

target("common2")
    ...

target("builder")
    add_deps("common1"， “common2")
    set_policy("build.across_targets_in_parallel", false)
    ...

target("run_builder")
    add_deps("builder")
    set_policy("build.across_targets_in_parallel", false)
    ...

target("app1")
    add_deps("run_builder")
    ...

target("app2")
    add_deps("run_builder")
    ...

```
根据依赖关系，我期望运行`xmake`的时候，编译顺序应该是 common1|common2 --> builder --> run_builder --> app1|app2
也就是两个common可以并行编译，builder和run_builder只能线性，最后编译两个app。
但是现在的结果是，common和app都并行编译了，好像并不符合add_deps文档里面的说明：
>添加当前目标的依赖目标，编译的时候，会去优先编译依赖的目标，然后再编译当前目标。。。

不是应该没有依赖关系的优先并行编译，有build.across_targets_in_parallel=false的单独编译，这样吗？

刚刚又试了一下，好像编译单个目标`xmake b app1`的时候是这样的顺序：
编译common1|common2，编译app1，生成common1|common2库，编译、链接builder，运行run_builder，链接app1

**Reply by waruqi** [reply]
across_targets_in_parallel 禁用的是所有 targets 之间的跨targets并行，你 app 上没禁用，app 和 common 里面的源文件就还是可以跨 targets 编译。。

**Reply by Domain** [reply]
这并行编译不是应该只在没有依赖关系的目标间进行吗？现在的做法是默认目标之间只存在链接关系，所以编译可以并行？那我这种只能全部线性才能正常？感觉可以改进一下

**Reply by waruqi** [reply]
> 没有依赖关系的目标间进行吗？现在的做法是默认目标之间只存在链接关系，所以编译可以并行？那我这种只能全部线性才能正常？感觉可以改进一下

不是，即使有 add_deps 依赖，通常大部分情况下，它们的所有 c/c++ 源文件之间还是没有任何依赖的，仅仅只有链接依赖。

所以这些源文件，完全可以最大化并行编译，跑满 cpu core，把所有 link 任务放到最后，对编译的速度提升是非常明显的，这也是 xmake 的一个重点优化方案。

如果想其他构建系统那样只能按 link 依赖来处理并行，那么一些只有几个 cpp 的 target，只能不断等 link 过后，才能编译他们，而这几个 cpp 内部并行也没啥明显优势了。。


所以，默认行为是源文件全部跨target 并行，只有一些特殊需求，源文件之间存在依赖关系的，才通过这个 policy 手动配置禁用，毕竟这种情况属于少数。

你这对 app 也设置下这个 policy ，或者干脆全局设置禁用下就行了。

而且即使全局设置，仅仅只是禁用了 跨 targets 间的源文件并行，但是 每个 target 内部的 cpp 文件，还是并行编译的，相当于退化到 其他构建系统常用的 并行方案而已。

并不是说，设置了这个 policy，所有 cpp 编译都线性编译了，只是 target 间根据 deps 关系线性处理，没 deps 关系的，还是 target 并行。

**Reply by Domain** [reply]
可是我有20+个app……这些都依赖于run_builder，很多链接时间都很长，如果都线性，效率显然低了很多。fastbuild没有这个问题，不知道它怎么处理这些依赖关系的

**Reply by waruqi** [reply]
> 可是我有20+个app……这些都依赖于run_builder，很多链接时间都很长，如果都线性，效率显然低了很多。fastbuild没有这个问题，不知道它怎么处理这些依赖关系的

那说明 fastbuild 可能都没这个targets间源文件并行优化，原本就是 target 间根据 deps 线性的。。

而且我刚说了，即使 targets 线性，target 内部 cpp 还是还是并行编译的。。而你既然有 autogen 导致的源文件之间依赖，那么 target 只能保持线性依赖。

**Reply by Domain** [reply]
不，fastbuild不单单不同target之间可以并行，甚至不同mode(debug, release, profile...)都可以并行，跑起来确实和我预期一致，可以并行的并行，不可以的线性，应该是仅仅根据依赖关系推算出来的
